crates/ty_python_semantic/resources/mdtest/type_properties/satisfied_by_all_typevars.md

```diff
@@ -0,0 +1,158 @@
+non-inferable.)
+
+## Inferable typevars
+
+When a typevar is in an inferable position, the constraint set only needs to be satisfied for _some_
+valid specialization. The most common inferable position occurs when invoking a generic function:
+all of the function's typevars are inferable, because we want to use the argument types to infer
+which specialization is being invoked.
+
+```py
+from typing import final, Never
+from ty_extensions import ConstraintSet, static_assert
+
+class Super: ...
+class Base(Super): ...
+class Sub(Base): ...
+
+@final
+class Unrelated: ...
+```
+
+If a typevar has no bound or constraints, then it can specialize to any type. For an inferable
+typevar, that means we just need a single type (any type at all!) that satisfies the constraint set.
+
+```py
+def unbounded[T]():
+    static_assert(ConstraintSet.always().satisfied_by_all_typevars(T))
+    static_assert(not ConstraintSet.never().satisfied_by_all_typevars(T))
+    static_assert(ConstraintSet.range(Never, T, Unrelated).satisfied_by_all_typevars(T))
+    static_assert(ConstraintSet.range(Never, T, Super).satisfied_by_all_typevars(T))
+    static_assert(ConstraintSet.range(Never, T, Base).satisfied_by_all_typevars(T))
+    static_assert(ConstraintSet.range(Never, T, Sub).satisfied_by_all_typevars(T))
+```
+
+If a typevar has an upper bound, then it must specialize to a type that is a subtype of that bound.
+For an inferable typevar, that means we need a single type that satisfies both the constraint set
+and the upper bound.
+
+```py
+def bounded[T: Base]():
+    static_assert(ConstraintSet.always().satisfied_by_all_typevars(T))
+    static_assert(not ConstraintSet.never().satisfied_by_all_typevars(T))
+    static_assert(ConstraintSet.range(Never, T, Super).satisfied_by_all_typevars(T))
+    static_assert(ConstraintSet.range(Never, T, Base).satisfied_by_all_typevars(T))
+    static_assert(ConstraintSet.range(Never, T, Sub).satisfied_by_all_typevars(T))
+
+    # This succeeds because T can specialize to Never
+    constraints = ConstraintSet.range(Never, T, Unrelated)
+    static_assert(constraints.satisfied_by_all_typevars(T))
+
+    # If we explicitly disallow Never, then it fails
+    constraints = constraints & ~ConstraintSet.range(Never, T, Never)
+    static_assert(not constraints.satisfied_by_all_typevars(T))
+```
+
+If a typevar has constraints, then it must specialize to one of those specific types. (Not to a
+subtype of one of those types!) For an inferable typevar, that means we need the constraint set to
+be satisfied by any one of the constraints.
+
+```py
+def constrained[T: (Base, Unrelated)]():
+    static_assert(ConstraintSet.always().satisfied_by_all_typevars(T))
+    static_assert(not ConstraintSet.never().satisfied_by_all_typevars(T))
+    static_assert(ConstraintSet.range(Never, T, Unrelated).satisfied_by_all_typevars(T))
+    static_assert(ConstraintSet.range(Never, T, Super).satisfied_by_all_typevars(T))
```

Comment:
Hmm, same here. Why is `Super` satisfying the constraint, given that it isn't `Base` or `Unrelated`?

Reply:
Same as above. This is in inferable position, so we only need one specialization of `T` to satisfy the constraint set. `T = Base` does, so the check passes. Down below in the non-inferable section, the check fails, because `T = Unrelated` is a valid specialization that doesn't satisfy the constraint set.